lib/src/main/java/org/kpipe/consumer/OffsetManager.java

```diff
@@ -0,0 +1,574 @@
+      if (result) {
+        cleanupCommittedOffsets(offsetsToCommit);
+      }
+      return result;
+    } catch (final ExecutionException | TimeoutException e) {
+      LOGGER.log(WARNING, "Error waiting for offset commit", e);
+      return false;
+    }
+  }
+
+  /**
+   * Cleans up offsets that have been successfully committed to prevent memory leaks.
+   *
+   * @param committedOffsets Map of offsets that were successfully committed
+   */
+  private void cleanupCommittedOffsets(final Map<TopicPartition, OffsetAndMetadata> committedOffsets) {
+    committedOffsets.forEach((partition, metadata) -> {
+      long committedOffset = metadata.offset();
+      final var processed = processedOffsets.get(partition);
+      if (processed != null) {
+        try {
+          // Get subSet view of elements less than committedOffset and clear it
+          final var toRemove = processed.headSet(committedOffset);
+          toRemove.clear();
+        } catch (final UnsupportedOperationException e) {
+          // Fall back to standard removeIf if headSet clear isn't supported
+          processed.removeIf(offset -> offset < committedOffset);
+        }
+
+        if (processed.isEmpty()) {
+          processedOffsets.remove(partition);
+        }
+      }
+    });
+  }
+
+  /**
+   * Updates the committable offset for a partition based on processed offsets. Handles both
+   * contiguous and non-contiguous offset tracking.
+   *
+   * @param partition The partition to update
+   */
+  public void updateCommittableOffset(final TopicPartition partition) {
+    final var currentPosition = nextOffsetsToCommit.get(partition);
+    if (currentPosition == null) {
+      return;
+    }
+
+    final var pending = pendingOffsets.get(partition);
+    final var processed = processedOffsets.get(partition);
+    if (processed == null || processed.isEmpty()) {
+      return;
+    }
+
+    // Find the highest contiguous processed offset
+    final var highestContiguous = findHighestContiguousOffset(processed);
+    final var nextExpected = highestContiguous + 1;
+
+    // If all offsets are processed, commit one past the highest contiguous
+    if (pending == null || pending.isEmpty()) {
+      nextOffsetsToCommit.put(partition, nextExpected);
+      return;
+    }
+
+    // If there are still pending offsets, find the lowest pending
+    // and commit up to that offset if there are no gaps before it
+    updateCommitPointBasedOnPendingOffsets(partition, currentPosition, pending, nextExpected);
+  }
+
+  /** Finds the highest contiguous offset in a set of processed offsets. */
+  private long findHighestContiguousOffset(final ConcurrentSkipListSet<Long> processed) {
+    var highestContiguous = processed.first();
+    var nextExpected = highestContiguous + 1;
+
+    // Find the highest contiguous sequence
+    while (processed.contains(nextExpected)) {
+      highestContiguous = nextExpected;
+      nextExpected++;
+    }
+
+    return highestContiguous;
+  }
```

Comment:
This wont always work. 
Offsets in Kafka can have gaps - they are not always contiguous - for example transactions create markers that occupy offsets but are not visible in consumer, log compaction will delete off offsets - so offsets are always increasing but not necessarily without gaps.
Changing logic as per my suggestion above will remove dependency on checking for gaps in processed offsets though.

Reply:
removed